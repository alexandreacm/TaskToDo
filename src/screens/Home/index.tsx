import React, {useCallback} from 'react';
import {
  SafeAreaView,
  StyleSheet,
  View,
  StatusBar,
  Platform,
} from 'react-native';

import {CustomHeader} from '../../components/CustomHeader';
import InputTask from '../../components/InputTask';
import TodoList from '../../components/TodoList';
import {clearStorage, containsKey, loadData, saveData} from '../../storage/';
import {TASK_KEY} from '../../constants';

import Button from '../../components/Button';
import {theme} from '../../styles/theme';

// const initialState = {
//   data: [],
//   newTask: "",
//   taskEditable: null,
//   error: false,
// };

//solution for android system.
const isAndroid = Platform.OS === 'android' ? StatusBar.currentHeight : 0;

function Home() {
  const [data, setData] = React.useState<Task[]>([]);
  const [newTask, setNewTask] = React.useState('');
  const [taskEditable, setTaskEditable] = React.useState(null);
  const [error, setError] = React.useState<boolean>(false);

  // console.log("Rendered...");

  React.useEffect(() => {
    // console.log("useEffect I");
    async function loadDataLocalStorage() {
      const isKeyTask = await containsKey(TASK_KEY);
      const localData = await loadData(TASK_KEY);

      if (isKeyTask && localData !== null) {
        setData(localData);
      }
    }

    loadDataLocalStorage();
  }, []);

  async function saveDataStorage() {
    if (data.length > 0) {
      await saveData(TASK_KEY, data);
    }
  }

  React.useEffect(() => {
    // console.log("useEffect II");
    saveDataStorage();
  }, [data]);

  const onHandleAddTask = () => {
    if (newTask === '') {
      setError(true);
      return;
    }

    const id = Math.floor(Math.random() * 100);
    const task: Task = {
      id,
      task: `${newTask}`,
      isCompleted: false,
    };

    setData([...data, task]);
    setNewTask('');
    setError(false);
  };

  const onHandleEditTask = () => {
    if (newTask === '') {
      setError(true);
      return;
    }

    const editingTask: any = data.find(item => item.id == taskEditable.idTask);
    editingTask.task = newTask;

    setData([...data]);

    setNewTask('');
    setError(false);
  };

  const onDeleteTask = (id: number) => {
    const taskItems = data.filter(task => task.id !== id);

    setData([...taskItems]);
  };

  const onEditTask = useCallback((item: Task) => {
    setNewTask(item.task);
    setTaskEditable({edit: true, idTask: item.id});
  }, []);

  const onFinishTask = (id: number) => {
    const taskItem: any = data.find(task => task.id == id);

    taskItem.isCompleted = !taskItem.isCompleted;
    setData([...data]);
  };

  return (
    <SafeAreaView style={styles.app}>
      <CustomHeader title={'Your Task'} />

      <TodoList
        data={data}
        onDeleteTask={onDeleteTask}
        onEditTask={onEditTask}
        onFinishTask={onFinishTask}
      />

      <CustomHeader title={'Add New Task'} />

      <View style={styles.bottomContainer}>
        <InputTask value={newTask} onChangeText={setNewTask} error={error} />

        <Button
          onPress={taskEditable ? onHandleEditTask : onHandleAddTask}
          text={taskEditable ? 'Edit' : 'Add'}
          style={buttonStyles.button}
          textStyle={buttonStyles.text}
          color={undefined}
          disabled={undefined}
        />
      </View>
    </SafeAreaView>
  );
}

const styles = StyleSheet.create({
  app: {
    flex: 1,
    backgroundColor: theme.colors.ui.WHITE,
    paddingTop: isAndroid,
  },
  bottomContainer: {
    padding: 16,
  },
});

const buttonStyles = StyleSheet.create({
  button: {
    width: '100%',
    height: 61,
    backgroundColor: theme.colors.ui.BUTTON,
    borderRadius: 30,
    justifyContent: 'center',
    alignItems: 'center',
    padding: 16,
    marginBottom: 25,
    marginTop: 55,
  },
  text: {
    color: theme.colors.text.WHITE,
    fontWeight: '700',
    fontSize: 16,
    textAlign: 'center',
    fontFamily: theme.fonts.Roboto_Regular,
  },
});

export default Home;
