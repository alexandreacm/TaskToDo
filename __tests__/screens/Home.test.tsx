import React from 'react';
import renderer, {act} from 'react-test-renderer';
import {render, fireEvent} from '@testing-library/react-native';
import Home from '../../src/screens/Home';

//Should check a finish task

describe('HOME', () => {
  jest.runAllTimers();

  const tree = renderer.create(<Home />).toJSON();

  test('Should render correctly component', () => {
    render(<Home />);
  });

  test('Should create a snapshot test correctly', () => {
    expect(tree).toMatchSnapshot();
  });

  it('Should test the Header title', () => {
    const {getByText} = render(<Home />);

    expect(getByText('Your Task')).toBeTruthy();
  });

  it('Should test the second Header title', () => {
    const {getByText} = render(<Home />);

    expect(getByText(' Add New Task')).toBeTruthy();
  });

  it('Should add a task', async () => {
    const {getByTestId, queryAllByText, debug} = render(<Home />);

    const inputTask = getByTestId('inputTask');
    const button = getByTestId('button');

    const createFirstTask = 'Task 1';

    act(() => {
      fireEvent.changeText(inputTask, createFirstTask);
      fireEvent.press(button);
    });

    const tasks = queryAllByText(/task/i);

    expect(tasks.length).toBe(2);
  });

  it('Should add multiple tasks', async () => {
    const {getByTestId, queryAllByText, debug} = render(<Home />);

    const createFirstTask_1 = 'Task 1';
    const createFirstTask_2 = 'Task 2';

    const inputTask = getByTestId('inputTask');
    const button = getByTestId('button');

    act(() => {
      fireEvent.changeText(inputTask, createFirstTask_1);
      fireEvent.press(button);

      fireEvent.changeText(inputTask, createFirstTask_2);
      fireEvent.press(button);
    });

    const tasks = queryAllByText(/task/i);

    expect(tasks.length).toBe(2);
  });

  it('Should validate if the field New Name is filled', () => {
    const {getByTestId} = render(<Home />);

    const button = getByTestId('button');
    fireEvent.press(button);

    const labelTaskError = getByTestId('labelTaskError');

    expect(labelTaskError).not.toBeNull();
  });

  it('Should validate if the error will appeared', () => {
    const {queryAllByText, getByTestId} = render(<Home />);

    const button = getByTestId('button');
    fireEvent.press(button);

    const labelTaskError = queryAllByText('*');

    expect(labelTaskError.length).toBe(1);
  });

  it('Should edit the task', () => {
    const {getByTestId, queryAllByText, queryByText, debug} = render(<Home />);

    const createFirstTask = 'first item';
    const createFirstTask_2 = 'second item';

    const inputTask = getByTestId('inputTask');
    const button = getByTestId('button');

    fireEvent.changeText(inputTask, createFirstTask);
    fireEvent.press(button);

    const onClickTask = queryAllByText(/first/i);
    fireEvent.press(onClickTask[0]);

    fireEvent.changeText(inputTask, createFirstTask_2);
    fireEvent.press(button);

    const taskRe = queryAllByText(/second/i);

    expect(taskRe.length).toBe(1);
  });

  it('should validate if Add button will be Edit button when edit a task', () => {
    const {getByTestId, queryAllByText, debug} = render(<Home />);

    const createFirstTask = 'first item';

    const inputTask = getByTestId('inputTask');
    const button = getByTestId('button');

    fireEvent.changeText(inputTask, createFirstTask);
    fireEvent.press(button);

    const onClickTask = queryAllByText(/first/i);
    fireEvent.press(onClickTask[0]);

    const editButton = queryAllByText(/edit/i);

    expect(editButton.length).toBe(1);
  });

  it('Should delete a task', () => {
    const {getByTestId, queryAllByText, debug} = render(<Home />);

    const createFirstItem = 'first item';
    const createSecondItem = 'second item';

    const inputTask = getByTestId('inputTask');
    const button = getByTestId('button');

    fireEvent.changeText(inputTask, createFirstItem);
    fireEvent.press(button);

    fireEvent.changeText(inputTask, createSecondItem);
    fireEvent.press(button);

    const removeButton = queryAllByText(/remove/i)[0];
    fireEvent.press(removeButton);

    expect(createSecondItem).not.toBe(null);
  });
});
