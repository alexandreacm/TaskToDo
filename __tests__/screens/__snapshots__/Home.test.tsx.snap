// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`HOME Should create a snapshot test correctly 1`] = `
<RCTSafeAreaView
  style={
    {
      "backgroundColor": "#FFFFFF",
      "flex": 1,
      "paddingTop": 0,
    }
  }
>
  <View
    style={
      {
        "alignItems": "center",
        "backgroundColor": "#D9EFFF",
        "height": 56,
        "justifyContent": "space-between",
        "marginTop": 10,
        "padding": 16,
        "width": "100%",
      }
    }
  >
    <Text
      style={
        {
          "color": "#0C8CE9",
          "fontFamily": "Inter_700Bold",
          "fontSize": 20,
          "fontWeight": "700",
          "height": 24,
          "lineHeight": 24,
          "textAlign": "center",
          "width": "100%",
        }
      }
      testID="textHeader"
    >
      Your Task
    </Text>
  </View>
  <View
    style={
      {
        "flex": 1,
        "padding": 16,
      }
    }
  >
    <RCTScrollView
      ItemSeparatorComponent={[Function]}
      data={[]}
      getItem={[Function]}
      getItemCount={[Function]}
      keyExtractor={[Function]}
      onContentSizeChange={[Function]}
      onLayout={[Function]}
      onMomentumScrollBegin={[Function]}
      onMomentumScrollEnd={[Function]}
      onScroll={[Function]}
      onScrollBeginDrag={[Function]}
      onScrollEndDrag={[Function]}
      removeClippedSubviews={false}
      renderItem={[Function]}
      scrollEventThrottle={0.0001}
      showVerticalScrollIndicator={false}
      stickyHeaderIndices={[]}
      testID="listTask"
      viewabilityConfigCallbackPairs={[]}
    >
      <View />
    </RCTScrollView>
  </View>
  <View
    style={
      {
        "alignItems": "center",
        "backgroundColor": "#D9EFFF",
        "height": 56,
        "justifyContent": "space-between",
        "marginTop": 10,
        "padding": 16,
        "width": "100%",
      }
    }
  >
    <Text
      style={
        {
          "color": "#0C8CE9",
          "fontFamily": "Inter_700Bold",
          "fontSize": 20,
          "fontWeight": "700",
          "height": 24,
          "lineHeight": 24,
          "textAlign": "center",
          "width": "100%",
        }
      }
      testID="textHeader"
    >
      Add New Task
    </Text>
  </View>
  <View
    style={
      {
        "padding": 16,
      }
    }
  >
    <View
      style={
        {
          "flexDirection": "row",
          "marginBottom": 7,
          "marginTop": 7,
        }
      }
    >
      <Text
        style={
          {
            "color": "#1E1E1E",
            "fontFamily": "Roboto_400Regular",
            "fontSize": 14,
            "fontWeight": "500",
            "lineHeight": 14.4,
            "marginRight": 5,
            "textAlign": "left",
          }
        }
        testID="labelTask"
      >
        Name
      </Text>
    </View>
    <View
      style={
        {
          "borderColor": "#666666",
          "borderRadius": 8,
          "borderWidth": 1,
          "height": 43,
          "width": "100%",
        }
      }
    >
      <TextInput
        onChangeText={[Function]}
        placeholder="New Name"
        placeholderTextColor="#666666"
        style={
          {
            "padding": 10,
          }
        }
        testID="inputTask"
        value="Task"
      />
    </View>
    <View
      accessibilityState={
        {
          "busy": undefined,
          "checked": undefined,
          "disabled": undefined,
          "expanded": undefined,
          "selected": undefined,
        }
      }
      accessibilityValue={
        {
          "max": undefined,
          "min": undefined,
          "now": undefined,
          "text": undefined,
        }
      }
      accessible={true}
      collapsable={false}
      focusable={true}
      onBlur={[Function]}
      onClick={[Function]}
      onFocus={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        [
          {
            "alignItems": "center",
            "alignSelf": "center",
            "backgroundColor": "#9747FF",
            "borderRadius": 50,
            "flexDirection": "row",
            "justifyContent": "center",
            "paddingHorizontal": 20,
            "paddingVertical": 8,
          },
          {
            "height": 40,
          },
          {
            "alignItems": "center",
            "backgroundColor": "#0C8CE9",
            "borderRadius": 30,
            "height": 61,
            "justifyContent": "center",
            "marginBottom": 25,
            "marginTop": 55,
            "padding": 16,
            "width": "100%",
          },
          {
            "opacity": 1,
          },
        ]
      }
      testID="button"
    >
      <Text
        style={
          [
            {
              "color": "#FFFFFF",
              "fontFamily": "Roboto_400Regular",
              "fontSize": 16,
              "fontWeight": "700",
              "textAlign": "center",
            },
            {},
          ]
        }
      >
        Add
      </Text>
    </View>
  </View>
</RCTSafeAreaView>
`;
